flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/ResultProvider.java

```diff
@@ -59,4 +59,7 @@ public interface ResultProvider {
      * {@link CloseableIterator#next()} method returns a row.
      */
     boolean isFirstRowReady();
+
+    /** Reset this ResultProvider to the origin state when we create it. */
+    default void reset() {}
```

Comment:
Adding a `reset` contract is fragile to break, I'm trying to understand the rational behind. If I'm not mistaken, the reason is `ResultProvider` is also cached, if we can make it bind only to a real job instance, then we'll don't need to reset it, do you think it's doable?

Reply:
`CollectResultProvider` used a field `iterator` of Outer class `CollectDynamicSink`, and it is **lazily created** in DataStreamSinkProvider#consumeDataStream(which is called when translate `DynamicTableSink` into a Transformation). So it is hard to recreate a new `CollectResultProvider` when we hit a plan cache as our caches store Transformations and will not create new Transformations.